### PR TITLE
avoid rpath'ing default search paths

### LINKED ISSUE #40848: Use of RPATH for implicit link directories problematic for build systems that use `LD_LIBRARY_PATH=<build dir> ./executable`
### Steps to reproduce this issue:

1. Use a compiler that has its runtime libraries in a directory that also contains `lib<example>.so`
2. Install the package `spack install <example>`
3. The build system invokes its just-built executable: `LD_LIBRARY_PATH=<build dir> ./example ...`

The latter fails because the `example` executable contains an RPATH pointing to the implicit link directory of the compiler, which happens to contain `libexample.so` -- this library is picked up instead of the just-built `libexample.so` in `LD_LIBRARY_PATH`.

Notice that Spack itself adds the rpath, the system compiler would only inject the `-L` flag.

The pattern `LD_LIBRARY_PATH=. ./example` is probably most common in autotools packages. CMake packages are not affected as they use build rpaths to work around this problem.

Below is an example where this happens with `perl`.

### Steps to reproduce the issue

```console
[csbcfmpd@login209-18 spack-src]$ spack spec -I perl
Input spec
--------------------------------
 -   perl

Concretized
--------------------------------
 -   perl@5.38.0%gcc@10.1.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-rocky8-icelake
[+]      ^berkeley-db@18.1.40%gcc@10.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rocky8-icelake
[+]          ^gmake@4.4.1%gcc@10.1.0~guile build_system=generic arch=linux-rocky8-icelake
[+]      ^bzip2@1.0.8%gcc@10.1.0~debug~pic+shared build_system=generic arch=linux-rocky8-icelake
[+]          ^diffutils@3.9%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]              ^libiconv@1.17%gcc@10.1.0 build_system=autotools libs=shared,static arch=linux-rocky8-icelake
[+]      ^gdbm@1.23%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]          ^readline@8.2%gcc@10.1.0 build_system=autotools patches=bbf97f1 arch=linux-rocky8-icelake
[+]              ^ncurses@6.4%gcc@10.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-rocky8-icelake
[+]                  ^pkgconf@1.9.5%gcc@10.1.0 build_system=autotools arch=linux-rocky8-icelake
[+]      ^zlib-ng@2.1.4%gcc@10.1.0+compat+opt build_system=autotools arch=linux-rocky8-icelake
```

### Error message

<details><summary>loadable library and perl binaries are mismatched</summary>
[csbcfmpd@login209-18 spack-src]$ spack install -j 40  perl
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/berkeley-db-18.1.40-i3j6vwjja62njh2mi6dgg5mqel3m6ani
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/bzip2-1.0.8-oo6spantejf27dmb7bzwccfttwtfx7bb
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/ncurses-6.4-2kenxblld7jmiylsjweggbentnz3yrbm
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/zlib-ng-2.1.4-575jc3ailnn2g7unxmoivbsov7vls7sw
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/readline-8.2-cpx6vq4aolpyqj7tguzpiha4hpuexnr5
[+] /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack_packages/linux-rocky8-icelake/gcc-10.1.0/gdbm-1.23-ei62qyttjgwe6qlz3sgbipffqt6b7ome
==> Installing perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 [7/7]
==> No binary for perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6 found: installing from source
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/21/213ef58089d2f2c972ea353517dc60ec3656f050dcc027666e118b508423e517.tar.gz
==> Using cached archive: /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/cache/_source-cache/archive/9d/9da50e155df72bce55cb69f51f1dbb4b62d23740fb99f6178bb27f22ebdf8a46.tar.gz
==> Moving resource stage
        source: /tmp/csbcfmpd/spack-stage/resource-cpanm-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/
        destination: /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/cpanm/cpanm
==> Applied patch /mnt/lustre/scratch/nlsas/home/csic/bcf/mpd/Simulation/spack_SeisSol/spack/var/spack/repos/builtin/packages/perl/zlib-ng.patch
==> Ran patch() for perl
==> perl: Executing phase: 'configure'
==> perl: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j40'

1 error found in build log:
     2831          \
     2832    
     2833    chmod 755 ../../lib/auto/Unicode/Normalize/Normalize.so
     2834    make[1]: Leaving directory '/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src/dist/Unicode-Normalize'
     2835    LD_LIBRARY_PATH=/tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src  ./perl -Ilib -I. -f pod/buildtoc -q
     2836    Cwd.c: loadable library and perl binaries are mismatched (got handshake key 0xf380080, needed 0xcd00080)
  >> 2837    make: *** [makefile:426: pod/perltoc.pod] Error 1

See build log for details:
  /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-build-out.txt
</details>

### Information on your system

* **Spack:** 0.21.0.dev0 (80944d22f775b014cfd201e895fc8777a2038786)
* **Python:** 3.7.8
* **Platform:** linux-rocky8-icelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/13244340/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/13244341/spack-build-out.txt)

The problem is that the perl version compiled in spack-src is not the 5.38.0 but the 5.30.3
```
[csbcfmpd@login209-18 ~]$ cd /tmp/csbcfmpd/spack-stage/spack-stage-perl-5.38.0-2tubazheh7m4pdei3eugarkkduubhga6/spack-src
[csbcfmpd@login209-18 spack-src]$ ./perl --version
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
        LANGUAGE = (unset),
        LC_ALL = (unset),
        LC_MEASUREMENT = "de_DE.UTF-8",
        LC_PAPER = "de_DE.UTF-8",
        LC_MONETARY = "de_DE.UTF-8",
        LC_NAME = "de_DE.UTF-8",
        LC_ADDRESS = "de_DE.UTF-8",
        LC_NUMERIC = "de_DE.UTF-8",
        LC_TELEPHONE = "de_DE.UTF-8",
        LC_IDENTIFICATION = "de_DE.UTF-8",
        LC_TIME = "de_DE.UTF-8",
        LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").

This is perl 5, version 30, subversion 3 (v5.30.3) built for x86_64-linux-thread-multi
(with 30 registered patches, see perl -V for more detail)

Copyright 1987-2020, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
```
maintainer: @LydDeb

The issue is related with:
https://github.com/spack/spack/issues/26738

A direct workaround is to install perl@5.30.3

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by haampie**
Closes #40848 

On sysroot systems like gentoo prefix, as well as nix/guix, our "is
system path" logic is broken cause it's static.

Talking about "the system paths" is not helpful, we have to talk
about default search paths of the dynamic linker instead.

If glibc is recent enough, we can query the dynamic loader's default
search paths, which is a much more robust way to avoid registering
rpaths to system dirs, which can shadow Spack dirs.

This PR adds an **additional** filter on rpaths the compiler wrapper
adds, dropping rpaths that are default search paths. The PR **does
not** remove any of the original `is_system_path` code, which is
questionable code, but can't hurt much.

(In a follow up PR we could look to register rpaths based *only* on
the default search paths of the dynamic linker, and get rid of
`is_system_path`. I don't want to make too many changes at the
same time.)

This fixes issues where build systems run just-built executables
linked against their *not-yet-installed libraries*, typically:

```
LD_LIBRARY_PATH=. ./exe
```

which happens in `perl`, `python`, and other non-cmake packages.
If a default path is rpath'ed, it takes precedence over
`LD_LIBRARY_PATH`, and a system library gets loaded instead
of the just-built library in the stage dir, breaking the build. If
default paths are not rpath'ed, then LD_LIBRARY_PATH takes
precedence, as is desired.

This PR additionally fixes an inconsistency in rpaths between
cmake and non-cmake packages. The cmake build system
computed rpaths by itself, but used a different order than
computed for the compiler wrapper. In fact it's not necessary
to compute rpaths at all, since we let cmake do that thanks to
`CMAKE_INSTALL_RPATH_USE_LINK_PATH`. This covers rpaths
for all dependencies. The only install rpaths we need to set are
`<install prefix>/{lib,lib64}`, which cmake unfortunately omits,
although it could also know these. Also, cmake does *not*
delete rpaths added by the toolchain (i.e. Spack's compiler
wrapper), so I don't think it should be controversial to simplify
things.

### CODE REVIEW

**Comment by haampie** [comment]
This still needs fixes for CMakePackage which doesn't follow the code path that filters the rpaths.

**Comment by haampie** [comment]
I realize that this may break cmake packages that do not hit the compiler wrapper due to `CMAKE_CXX_COMPILER=hipcc`. `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON` is sufficient to register rpaths to non-project libraries, but will unfortunately not set an rpath to the package's own install prefix. The compiler wrapper registers `<prefix>/lib[64]`.

**Comment by Chrismarsh** [comment]
I rebased this PR on develop and I'm seeing the following build error
```
==> Installing cmake-3.30.2-ylbdewin44le3plnp77jd2paeodaacpk [73/209]
==> No binary for cmake-3.30.2-ylbdewin44le3plnp77jd2paeodaacpk found: installing from source
==> Fetching https://github.com/Kitware/CMake/releases/download/v3.30.2/cmake-3.30.2.tar.gz
==> Applied patch /globalhome/cbm038/HPC/code/spack/var/spack/repos/builtin/packages/cmake/mr-9623.patch
==> cmake: Executing phase: 'bootstrap'
==> Error: AttributeError: module 'spack.build_environment' has no attribute 'get_rpaths'

The 'cmake' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

/globalhome/cbm038/HPC/code/spack/var/spack/repos/builtin/packages/cmake/package.py:336, in bootstrap_args:
        333        )
        334
        335        # Make CMake find its own dependencies.
  >>    336        rpaths = spack.build_environment.get_rpaths(self)
        337        prefixes = spack.build_environment.get_cmake_prefix_path(self)
        338        args.extend(
        339            [

See build log for details:
```

**Comment by Chrismarsh** [comment]
So far this PR seems to be working well for me on gentoo. 

This change https://github.com/spack/spack/commit/c4dd67b7a54eb5f6e80c199c75e5407438c6200d got me back up and running after I rebased this on a more recent develop

What can I test / provide that can help push this PR forward? Being able to use spack with ease on gentoo would let me ditch `easy_build`

**Comment by haampie** [comment]
OK, so the CMake solution of the PR is to use `CMAKE_INSTALL_RPATH` only for `<prefix>/{lib,lib64}`, and rely for the rest on the compiler wrapper and `CMAKE_INSTALL_RPATH_USE_LINK_PATH` to set rpaths to dependencies.

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -435,6 +437,35 @@ def optimization_flags(compiler, target):
     return result
 
 
+class FilterDefaultDynamicLinkerSearchPaths:
+    """Remove rpaths to directories that are default search paths of the dynamic linker."""
+
+    def __init__(self, dynamic_linker: Optional[str]) -> None:
+        # Idenitify directories by (inode, device) tuple, which handles symlinks too.
```
Comment (alalazo) [inline lib/spack/spack/build_environment.py]:
```suggestion
        # Identify directories by (inode, device) tuple, which handles symlinks too.
```

**Review by alalazo — DISMISSED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**